File: plataforma/backend/src/services/project.service.ts
/**
 * Project Service
 * Maneja la lógica de negocio para proyectos finales
 * Entrega, revisión y calificación de proyectos
 */

import { prisma } from '../utils/prisma';
import { NotFoundError, ValidationError } from '../middleware/errorHandler';
import { logger } from '../middleware/logger';
import notificationService from './notification.service';

/**
 * Interfaz para enviar proyecto
 */
export interface SubmitProjectData {
  projectId: string;
  userId: string;
  repositoryUrl?: string;
  files?: Array<{ name: string; url: string; size: number }>;
  description?: string;
}

/**
 * Interfaz para calificar proyecto
 */
export interface GradeProjectData {
  submissionId: string;
  instructorId: string;
  score: number;
  feedback: string;
  status: 'APPROVED' | 'REJECTED';
}

/**
 * Project Service
 * Proporciona operaciones para gestión de proyectos finales
 */
class ProjectService {
  /**
   * Obtener proyecto por ID
   * @param projectId - ID del proyecto
   * @param userId - ID del usuario (para verificar enrollment)
   * @returns Proyecto con información completa
   */
  async getProjectById(projectId: string, userId: string) {
    const project = await prisma.project.findUnique({
      where: { id: projectId },
      include: {
        course: {
          select: {
            id: true,
            slug: true,
            title: true,
          },
        },
      },
    });

    if (!project) {
      throw new NotFoundError('Proyecto no encontrado');
    }

    // Verificar que el usuario está inscrito en el curso
    const enrollment = await prisma.enrollment.findUnique({
      where: {
        userId_courseId: {
          userId,
          courseId: project.courseId,
        },
      },
    });

    if (!enrollment) {
      throw new ValidationError('No estás inscrito en este curso');
    }

    // Obtener entregas del usuario
    const userSubmissions = await prisma.projectSubmission.findMany({
      where: {
        projectId,
        userId,
      },
      orderBy: {
        submittedAt: 'desc',
      },
      select: {
        id: true,
        status: true,
        score: true,
        feedback: true,
        submittedAt: true,
        reviewedAt: true,
      },
    });

    return {
      ...project,
      userSubmissions,
    };
  }

  /**
   * Enviar proyecto
   * @param data - Datos del proyecto
   * @returns Submission creado
   */
  async submitProject(data: SubmitProjectData) {
    // Verificar que el proyecto existe
    const project = await prisma.project.findUnique({
      where: { id: data.projectId },
      include: {
        course: {
          select: {
            title: true,
          },
        },
      },
    });

    if (!project) {
      throw new NotFoundError('Proyecto no encontrado');
    }

    // Verificar enrollment
    const enrollment = await prisma.enrollment.findUnique({
      where: {
        userId_courseId: {
          userId: data.userId,
          courseId: project.courseId,
        },
      },
    });

    if (!enrollment) {
      throw new ValidationError('No estás inscrito en este curso');
    }

    // Verificar si ya tiene una entrega pendiente
    const pendingSubmission = await prisma.projectSubmission.findFirst({
      where: {
        projectId: data.projectId,
        userId: data.userId,
        status: { in: ['PENDING', 'REVIEWING'] },
      },
    });

    if (pendingSubmission) {
      throw new ValidationError('Ya tienes una entrega en revisión para este proyecto');
    }

    // Crear submission
    const submission = await prisma.projectSubmission.create({
      data: {
        projectId: data.projectId,
        userId: data.userId,
        repositoryUrl: data.repositoryUrl,
        files: data.files || null,
        description: data.description,
        status: 'PENDING',
      },
      include: {
        project: {
          select: {
            title: true,
          },
        },
        user: {
          select: {
            name: true,
            email: true,
          },
        },
      },
    });

    logger.info(
      `Proyecto "${submission.project.title}" enviado por ${submission.user.email}`
    );

    return submission;
  }

  /**
   * Obtener submissions pendientes de revisión (para instructores)
   * @param instructorId - ID del instructor
   * @param courseId - ID del curso (opcional)
   * @returns Lista de submissions pendientes
   */
  async getPendingSubmissions(instructorId: string, courseId?: string) {
    // Verificar que el usuario es INSTRUCTOR o ADMIN
    const instructor = await prisma.user.findUnique({
      where: { id: instructorId },
      select: { role: true },
    });

    if (!instructor || !['INSTRUCTOR', 'ADMIN'].includes(instructor.role)) {
      throw new ValidationError('Solo instructores pueden revisar proyectos');
    }

    const where: any = {
      status: { in: ['PENDING', 'REVIEWING'] },
    };

    if (courseId) {
      where.project = {
        courseId,
      };
    }

    const submissions = await prisma.projectSubmission.findMany({
      where,
      include: {
        user: {
          select: {
            id: true,
            name: true,
            email: true,
            avatar: true,
          },
        },
        project: {
          select: {
            id: true,
            title: true,
            courseId: true,
            course: {
              select: {
                title: true,
                slug: true,
              },
            },
          },
        },
      },
      orderBy: {
        submittedAt: 'asc',
      },
    });

    return submissions;
  }

  /**
   * Calificar proyecto
   * @param data - Datos de la calificación
   * @returns Submission calificado
   */
  async gradeProject(data: GradeProjectData) {
    // Verificar que el submission existe
    const submission = await prisma.projectSubmission.findUnique({
      where: { id: data.submissionId },
      include: {
        user: {
          select: {
            id: true,
            name: true,
            email: true,
          },
        },
        project: {
          select: {
            title: true,
          },
        },
      },
    });

    if (!submission) {
      throw new NotFoundError('Entrega de proyecto no encontrada');
    }

    // Verificar que el instructor tiene permisos
    const instructor = await prisma.user.findUnique({
      where: { id: data.instructorId },
      select: { role: true, name: true },
    });

    if (!instructor || !['INSTRUCTOR', 'ADMIN'].includes(instructor.role)) {
      throw new ValidationError('Solo instructores pueden calificar proyectos');
    }

    // Validar score
    if (data.score < 0 || data.score > 100) {
      throw new ValidationError('La calificación debe estar entre 0 y 100');
    }

    // Actualizar submission
    const updated = await prisma.projectSubmission.update({
      where: { id: data.submissionId },
      data: {
        status: data.status,
        score: data.score,
        feedback: data.feedback,
        reviewedBy: data.instructorId,
        reviewedAt: new Date(),
      },
      include: {
        user: {
          select: {
            id: true,
            name: true,
            email: true,
          },
        },
        project: {
          select: {
            title: true,
          },
        },
      },
    });

    logger.info(
      `Proyecto "${updated.project.title}" de ${updated.user.email} calificado: ${data.score}% - ${data.status}`
    );

    // Crear notificación
    await notificationService.notifyProjectGraded(
      submission.user.id,
      submission.project.title,
      data.score,
      data.status
    );

    return updated;
  }

  /**
   * Obtener submissions de un usuario
   * @param userId - ID del usuario
   * @param courseId - ID del curso (opcional)
   * @returns Lista de submissions del usuario
   */
  async getUserSubmissions(userId: string, courseId?: string) {
    const where: any = {
      userId,
    };

    if (courseId) {
      where.project = {
        courseId,
      };
    }

    const submissions = await prisma.projectSubmission.findMany({
      where,
      include: {
        project: {
          select: {
            id: true,
            title: true,
            course: {
              select: {
                id: true,
                title: true,
                slug: true,
              },
            },
          },
        },
        reviewer: {
          select: {
            name: true,
          },
        },
      },
      orderBy: {
        submittedAt: 'desc',
      },
    });

    return submissions;
  }

  /**
   * Obtener detalle de submission
   * @param submissionId - ID del submission
   * @param userId - ID del usuario (para verificar permisos)
   * @returns Submission completo
   */
  async getSubmissionDetail(submissionId: string, userId: string) {
    const submission = await prisma.projectSubmission.findUnique({
      where: { id: submissionId },
      include: {
        user: {
          select: {
            id: true,
            name: true,
            email: true,
            avatar: true,
          },
        },
        project: {
          include: {
            course: {
              select: {
                id: true,
                title: true,
                slug: true,
              },
            },
          },
        },
        reviewer: {
          select: {
            name: true,
            email: true,
          },
        },
      },
    });

    if (!submission) {
      throw new NotFoundError('Entrega de proyecto no encontrada');
    }

    // Verificar permisos: debe ser el autor o un instructor/admin
    const user = await prisma.user.findUnique({
      where: { id: userId },
      select: { role: true },
    });

    if (submission.userId !== userId && !['INSTRUCTOR', 'ADMIN'].includes(user?.role || '')) {
      throw new ValidationError('No tienes permiso para ver esta entrega');
    }

    return submission;
  }
}

// Exportar instancia singleton
export default new ProjectService();

File: plataforma/backend-fastapi/app/services/project_service.py
import structlog
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.middleware.error_handler import NotFoundError, ValidationError
from app.models.assessment import Project, ProjectStatus, ProjectSubmission

logger = structlog.get_logger()


class ProjectService:
    def __init__(self, db: AsyncSession):
        self.db = db

    async def get_by_id(self, project_id: str) -> Project:
        result = await self.db.execute(select(Project).where(Project.id == project_id))
        project = result.scalar_one_or_none()
        if not project:
            raise NotFoundError("Proyecto no encontrado")
        return project

    async def list_by_module(self, module_id: str) -> list[Project]:
        result = await self.db.execute(
            select(Project).where(Project.module_id == module_id).order_by(Project.created_at)
        )
        return list(result.scalars().all())

    async def create(self, module_id: str, data: dict) -> Project:
        project = Project(module_id=module_id, **data)
        self.db.add(project)
        await self.db.flush()
        logger.info("project_created", project_id=project.id, module_id=module_id)
        return project

    async def update(self, project_id: str, data: dict) -> Project:
        project = await self.get_by_id(project_id)
        for key, value in data.items():
            if value is not None and hasattr(project, key):
                setattr(project, key, value)
        await self.db.flush()
        return project

    async def delete(self, project_id: str) -> None:
        project = await self.get_by_id(project_id)
        await self.db.delete(project)
        await self.db.flush()
        logger.info("project_deleted", project_id=project_id)

    async def submit(self, project_id: str, user_id: str, content: str | None = None, file_url: str | None = None) -> ProjectSubmission:
        await self.get_by_id(project_id)
        submission = ProjectSubmission(
            user_id=user_id,
            project_id=project_id,
            content=content,
            file_url=file_url,
            status=ProjectStatus.PENDING,
        )
        self.db.add(submission)
        await self.db.flush()
        logger.info("project_submitted", project_id=project_id, user_id=user_id)
        return submission

    async def get_submission(self, submission_id: str) -> ProjectSubmission:
        result = await self.db.execute(select(ProjectSubmission).where(ProjectSubmission.id == submission_id))
        sub = result.scalar_one_or_none()
        if not sub:
            raise NotFoundError("Entrega no encontrada")
        return sub

    async def list_submissions(self, project_id: str | None = None, user_id: str | None = None, status: str | None = None) -> list[ProjectSubmission]:
        query = select(ProjectSubmission)
        if project_id:
            query = query.where(ProjectSubmission.project_id == project_id)
        if user_id:
            query = query.where(ProjectSubmission.user_id == user_id)
        if status:
            query = query.where(ProjectSubmission.status == ProjectStatus(status))
        query = query.order_by(ProjectSubmission.created_at.desc())
        result = await self.db.execute(query)
        return list(result.scalars().all())

    async def review_submission(self, submission_id: str, reviewer_id: str, status: str, score: float | None = None, feedback: str | None = None) -> ProjectSubmission:
        sub = await self.get_submission(submission_id)
        if sub.status not in (ProjectStatus.PENDING, ProjectStatus.REVIEWING):
            raise ValidationError("Esta entrega ya fue revisada")

        sub.status = ProjectStatus(status)
        sub.reviewer_id = reviewer_id
        sub.score = score
        sub.feedback = feedback
        await self.db.flush()

        if sub.status == ProjectStatus.APPROVED:
            project = await self.get_by_id(sub.project_id)
            from app.services.progress_service import ProgressService
            ps = ProgressService(self.db)
            await ps.update_project_status(sub.user_id, project.module_id, "APPROVED")

        logger.info("submission_reviewed", submission_id=submission_id, status=status)
        return sub
